hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetPlan.java

```diff
@@ -111,216 +243,386 @@ boolean ifExists() {
             return ifExists;
         }
 
+        @Override
+        public void checkPermissions(SqlSecurityContext context) {
+        }
+
         @Override
         public boolean producesRows() {
             return false;
         }
+
+        @Override
+        public boolean isPlanValid(PlanCheckContext context) {
+            return true;
+        }
+
+        @Override
+        public SqlResult execute(QueryId queryId) {
+            return planExecutor.execute(this);
+        }
     }
 
-    class CreateJobPlan implements JetPlan {
-        private final String jobName;
+    interface CreateJobPlan extends JetPlan {
+        JobConfig getJobConfig();
+
+        boolean isIfNotExists();
+
+        SelectOrSinkPlan getExecutionPlan();
+    }
+
+    class NonCacheableCreateJobPlan implements CreateJobPlan {
         private final JobConfig jobConfig;
         private final boolean ifNotExists;
         private final SelectOrSinkPlan dmlPlan;
         private final JetPlanExecutor planExecutor;
 
-        CreateJobPlan(
-                String jobName,
+        private NonCacheableCreateJobPlan(
                 JobConfig jobConfig,
                 boolean ifNotExists,
                 SelectOrSinkPlan dmlPlan,
                 JetPlanExecutor planExecutor
         ) {
-            this.jobName = jobName;
             this.jobConfig = jobConfig;
             this.ifNotExists = ifNotExists;
             this.dmlPlan = dmlPlan;
             this.planExecutor = planExecutor;
         }
 
+        @Override
+        public JobConfig getJobConfig() {
+            return jobConfig;
+        }
+
+        @Override
+        public boolean isIfNotExists() {
+            return ifNotExists;
+        }
+
+        @Override
+        public SelectOrSinkPlan getExecutionPlan() {
+            return dmlPlan;
+        }
+
         @Override
         public void checkPermissions(SqlSecurityContext context) {
             dmlPlan.checkPermissions(context);
         }
 
+        @Override
+        public boolean producesRows() {
+            return false;
+        }
+
         @Override
         public SqlResult execute(QueryId queryId) {
             return planExecutor.execute(this);
         }
+    }
 
-        public String getJobName() {
-            return jobName;
+    class CacheableCreateJobPlan extends CacheableJetPlan implements CreateJobPlan {
```

Comment:
This looks like combinatorial explosion of class hierarchy. Maybe we could merge the `CacheablePlan` interface into `SqlPlan` and add a method `isCacheable`. It can maybe even be an abstract class and implement the `onPlanUsed` and `getPlanLastUsed` methods. Then this class and the `NonCacheableCreateJobPlan` can be merged.

Reply:
Totally agree, wanted to do exactly that refactoring in a separate PR.

Reply:
OK